nemo_curator/modules/semdedup/semdedup.py

```diff
@@ -0,0 +1,207 @@
+    del cluster_reps
+    pair_w_sim_matrix.fill_diagonal_(0.0)
+    assert pair_w_sim_matrix.shape[0] == pair_w_sim_matrix.shape[1]
+
+    triu_sim_mat = torch.triu(pair_w_sim_matrix, diagonal=1)
+
+    M = torch.max(triu_sim_mat, dim=0)[0].cpu()
+    M1 = torch.max(triu_sim_mat, dim=0)[1].cpu().numpy().tolist()
+    return M, M1
+
+
+def get_cluster_reps(
+    cluster_id: int, emb_by_clust_loc: str, id_col: str, sorted_ids: np.ndarray
+) -> torch.Tensor:
+    cluster_i_path = os.path.join(emb_by_clust_loc, f"nearest_cent={cluster_id}")
+    cluster_reps = cudf.read_parquet(
+        cluster_i_path, columns=["embeddings", id_col]
+    ).sort_values(by=id_col)
+    num = cluster_reps.shape[0]
+
+    df_ = pd.DataFrame(
+        {"sorted_ids": sorted_ids, "inverse_sort": list(range(num))}
+    ).sort_values(by="sorted_ids")
+    cluster_reps["inverse_sort_id"] = df_["inverse_sort"].values
+    cluster_reps = cluster_reps.sort_values(by="inverse_sort_id")
+
+    cluster_reps = torch.as_tensor(
+        cluster_reps["embeddings"].list.leaves.values.reshape(len(cluster_reps), -1),
+        device="cuda",
+    )
+    return cluster_reps
+
+
+def process_cluster(
+    cluster_id: int,
+    emb_by_clust_loc: str,
+    id_col: str,
+    id_col_type: str,
+    eps_list: List[float],
+    save_loc: str,
+) -> None:
+    assert save_loc is not None
+    assert save_loc != ""
+
+    df_file_loc = os.path.join(save_loc, f"dataframes/cluster_{cluster_id}.parquet")
+    if os.path.exists(df_file_loc):
+        logging.info(f"{df_file_loc} exists. Continue")
+        return
+
+    sorted_clusters_path = f"{save_loc}/sorted"
+    sorted_file = os.path.join(sorted_clusters_path, f"cluster_{cluster_id}.npy")
+    if not os.path.exists(sorted_file):
+        logging.info(f"{sorted_file} does not exist. Continue")
+        return
+
+    cluster_i = np.load(sorted_file)
+
+    cluster_size = cluster_i.shape[0]
+    logging.info(f"{cluster_id}: cluster_size: {cluster_size}")
+
+    if cluster_size == 1:
+        points_to_remove_df = pd.DataFrame()
+        points_to_remove_df["indices"] = [0]
+        for eps in eps_list:
+            points_to_remove_df[f"eps={eps}"] = [False]
+        points_to_remove_df.to_parquet(df_file_loc)
+        return
+
+    clutser_items_indices = list(range(cluster_size))
+
+    which_to_keep = args.semdedup["which_to_keep"].lower()
+    if which_to_keep == "random":
+        random.shuffle(clutser_items_indices)
+        cluster_i = cluster_i[clutser_items_indices]
+    elif which_to_keep == "easy":
+        clutser_items_indices = clutser_items_indices[::-1]
+        cluster_i = cluster_i[clutser_items_indices]
+
+    text_ids = cluster_i[:, 0].astype(id_col_type)
+
+    cluster_reps = get_cluster_reps(cluster_id, emb_by_clust_loc, id_col, text_ids)
+    M, M1 = _semdedup(cluster_reps, "cuda")
+    assert cluster_reps.shape[0] == len(text_ids)
+
+    # TODO: Below is not used , ask @Fay if we can remove it
```

Comment:
This is for debugging/validation purposes. M returns the max cos-sim score for every text.  M1 returns the corresponding text which has the max cos-sim score with that text.  By keeping the M1 information, we will be able to cross check how similar they are.